libraries/ms-drivers/src/stm32f0xx/mcp3427_adc.c

```diff
@@ -0,0 +1,180 @@
+#include "mcp3427_adc.h"
+#include "fsm.h"
+#include "log.h"
+#include "mcp3427_adc_defs.h"
+#include "soft_timer.h"
+
+// Note: we always read from both channels before reporting, taking 100ms between reports. If timing
+// is an issue, an optimization for boards (i.e. solar) using only one channel is to only read that
+// channel, so reports take 50ms.
+
+// We use a combination of the address and I2C port in the event data to only transition on events
+// that we raised. Possible optimization if necessary: make prv_get_chip_identifier public, then
+// keep a reverse lookup table of identifier -> MCP3427 storage and only call the appropriate
+// MCP3427 when receiving an MCP3427 event. That way, using n MCP3427s generates n calls to
+// |mcp3427_process_event| per cycle rather than n^2 if we pass every event to every MCP3427.
+// (I think this was done on MSXII.)
```

Comment:
I think this might be worth implementing since we can have a theoretical maximum of 18 MCP3427s, meaning 324 calls to `mcp3427_process_event` each cycle. I'll implement it.